openwhisk/openwhisk_apigateway.md

```diff
@@ -15,58 +15,111 @@ lastupdated: "2017-03-16"
+
+## Expose an OpenWhisk web action
+
+1. Create a JavaScript file with the following content. For this example, the file name is 'hello.js'.
+  ```javascript
+  function main({name:name='Serverless API'}) {
+      return {payload: `Hello world ${name}`};
+  }
+  ```
+  {: codeblock}
+  
+2. Create a web action from the following JavaScript function. For this example, the action is called 'hello'. Make sure to add the flag `--web true`
+  
+  ```
+  wsk action update hello hello.js --web true
+  ```
+  {: pre}
+  ```
+  ok: updated action hello
+  ```
+  
+3. Create an API with base path `/hello`, path `/world` and method `get` with response type `json`
+  
+  ```
+  wsk api create /hello /world get hello --response-type json
+  ```
+  {: pre}
+  ```
+  ok: created API /hello/world GET for action /_/hello
+  https://service.us.apiconnect.ibmcloud.com/gws/apigateway/api/21ef035/hello/world
+  ```
+  A new URL is generated exposing the `hello` action via a __GET__ HTTP method.
+  
+4. Let's give it a try by sending a HTTP request to the URL.
+  
+  ```
+  $ curl https://service.us.apiconnect.ibmcloud.com/gws/apigateway/api/21ef035/hello/world?name=OpenWhisk
+  ```
+  {: pre}
+  ```json
+  {
+  "payload": "Hello world OpenWhisk"
+  }
+  ```
+  The action `hello` got invoked, returning back a JSON string including the parameter `name` sent via query parameter. You can pass parameters to the action via simple query parameters, or via request body.
+  
+5. Full controll over the the http response
+  
+  Notice that the full result of the action is returned in the body of the response, and the content is assume to be `application/json` this is because the API was created with `--response-type json` using `json` for the response-type allows you to easily get started with existing actions, but once you have something working you want to have full control over the http response like `statusCode`, `headers` and return different content types in the `body`.
+  
+  You can choose to change the code of the action to comply with the return of web actions with http extension or include the action in a sequence passing it's result to a new action that transform the result to be properly formatted for an http response. You can read more about response types and web actions extensions in the [Web Actions](webactions.md) documentation.
+
+  Change the code for the `hello.js` returning the JSON properties `body`, `statusCode` and `headers`
+  ```javascript
+  function main({name:name='Serverless API'}) {
+      return {
+        body: new Buffer(JSON.stringify({payload:`Hello world ${name}`})).toString('base64'), 
+        statusCode:200, 
+        headers:{ 'Content-Type': 'application/json'}
+      };
+  }
+  ```
+  Notice that the body needs to be return encoded in `base64` and not a string.
```

Comment:
...that the `JSON` body need to be returned `as base64 encoded` and not a string